File: java/graphs/BellmanFordProduction.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class BellmanFordProduction {
    record Cell(int row, int col) {}

    public static List<Cell> bellmanFordPath(
            boolean[][] walls,
            int[][] costs,
            Cell start,
            Cell target
    ) {
        int height = walls.length;
        int width = walls[0].length;
        validateCell(start, width, height);
        validateCell(target, width, height);

        if (walls[start.row()][start.col()] || walls[target.row()][target.col()]) {
            return List.of();
        }

        int total = width * height;
        int startId = id(start, width);
        int targetId = id(target, width);
        int[] distance = new int[total];
        int[] parent = new int[total];
        java.util.Arrays.fill(distance, Integer.MAX_VALUE);
        java.util.Arrays.fill(parent, -1);
        distance[startId] = 0;

        for (int pass = 1; pass < total; pass++) {
            boolean changed = false;

            for (int currentId = 0; currentId < total; currentId++) {
                if (walls[currentId / width][currentId % width]
                        || distance[currentId] == Integer.MAX_VALUE) {
                    continue;
                }

                int row = currentId / width;
                int col = currentId % width;
                int[][] moves = {
                    {-1, 0},
                    {0, 1},
                    {1, 0},
                    {0, -1}
                };

                for (int[] move : moves) {
                    int nextRow = row + move[0];
                    int nextCol = col + move[1];
                    if (nextRow < 0 || nextRow >= height || nextCol < 0 || nextCol >= width
                            || walls[nextRow][nextCol]) {
                        continue;
                    }

                    int nextId = nextRow * width + nextCol;
                    int candidate = distance[currentId] + costs[nextRow][nextCol];
                    if (candidate < distance[nextId]) {
                        distance[nextId] = candidate;
                        parent[nextId] = currentId;
                        changed = true;
                    }
                }
            }

            if (!changed) {
                break;
            }
        }

        for (int currentId = 0; currentId < total; currentId++) {
            int row = currentId / width;
            int col = currentId % width;
            if (walls[row][col] || distance[currentId] == Integer.MAX_VALUE) {
                continue;
            }

            int[][] moves = {
                {-1, 0},
                {0, 1},
                {1, 0},
                {0, -1}
            };

            for (int[] move : moves) {
                int nextRow = row + move[0];
                int nextCol = col + move[1];
                if (nextRow >= 0 && nextRow < height && nextCol >= 0 && nextCol < width
                        && !walls[nextRow][nextCol]
                        && distance[currentId] + costs[nextRow][nextCol]
                        < distance[nextRow * width + nextCol]) {
                    throw new IllegalStateException("reachable negative cycle");
                }
            }
        }

        if (distance[targetId] == Integer.MAX_VALUE) {
            return List.of();
        }

        List<Cell> path = new ArrayList<>();
        for (int currentId = targetId; currentId != -1; currentId = parent[currentId]) {
            path.add(new Cell(currentId / width, currentId % width));
        }
        Collections.reverse(path);
        return path;
    }

    private static int id(Cell cell, int width) {
        return cell.row() * width + cell.col();
    }

    private static void validateCell(Cell cell, int width, int height) {
        if (cell.row() < 0 || cell.row() >= height || cell.col() < 0 || cell.col() >= width) {
            throw new IllegalArgumentException("start and target must be inside the grid");
        }
    }
}

File: python/graphs/bellman_ford_production.py
from collections.abc import Iterable, Mapping


Cell = tuple[int, int]


def bellman_ford_path(
    width: int,
    height: int,
    walls: Iterable[Cell],
    costs: Mapping[Cell, int],
    start: Cell,
    target: Cell,
) -> list[Cell]:
    if width <= 0 or height <= 0:
        raise ValueError("grid dimensions must be positive")

    def inside(cell: Cell) -> bool:
        return 0 <= cell[0] < height and 0 <= cell[1] < width

    if not inside(start) or not inside(target):
        raise ValueError("start and target must be inside the grid")

    total = width * height
    blocked = bytearray(total)
    for wall in walls:
        if inside(wall):
            blocked[wall[0] * width + wall[1]] = 1

    start_id = start[0] * width + start[1]
    target_id = target[0] * width + target[1]
    if blocked[start_id] or blocked[target_id]:
        return []

    cells = [cell_id for cell_id in range(total) if not blocked[cell_id]]
    infinity = total * max(1, max((abs(value) for value in costs.values()), default=1)) + 1
    distance = [infinity] * total
    parent = [-1] * total
    distance[start_id] = 0

    def neighbors(cell_id: int):
        row, col = divmod(cell_id, width)
        for next_row, next_col in (
            (row - 1, col),
            (row, col + 1),
            (row + 1, col),
            (row, col - 1),
        ):
            if 0 <= next_row < height and 0 <= next_col < width:
                next_id = next_row * width + next_col
                if not blocked[next_id]:
                    yield next_id

    for _ in range(len(cells) - 1):
        changed = False

        for cell_id in cells:
            if distance[cell_id] == infinity:
                continue

            for next_id in neighbors(cell_id):
                next_cell = divmod(next_id, width)
                candidate = distance[cell_id] + costs.get(next_cell, 1)
                if candidate >= distance[next_id]:
                    continue

                distance[next_id] = candidate
                parent[next_id] = cell_id
                changed = True

        if not changed:
            break

    for cell_id in cells:
        if distance[cell_id] == infinity:
            continue
        for next_id in neighbors(cell_id):
            next_cell = divmod(next_id, width)
            if distance[cell_id] + costs.get(next_cell, 1) < distance[next_id]:
                raise ValueError("reachable negative cycle")

    if distance[target_id] == infinity:
        return []

    path = []
    current_id = target_id
    while current_id != -1:
        path.append(divmod(current_id, width))
        current_id = parent[current_id]
    path.reverse()
    return path
